### PR TITLE
Check datetime instantiation

### PR DESCRIPTION
**Pull request by b1rdex**
Fixes https://github.com/phpstan/phpstan/issues/4096

### CODE REVIEW

File: conf/config.level0.neon [hunk 1]
```diff
@@ -55,6 +55,7 @@ rules:
 	- PHPStan\Rules\Properties\AccessStaticPropertiesInAssignRule
 	- PHPStan\Rules\Properties\PropertyAttributesRule
 	- PHPStan\Rules\Variables\UnsetRule
+	- PHPStan\Rules\DateTimeInstantiationRule
```
Comment (ondrejmirtes) [inline conf/config.level0.neon:58]: 1) Needs to be on level 5 to be consistent with other rules
2) Needs to be only in bleeding edge. See an example of how to set it up: https://github.com/phpstan/phpstan-src/blob/41ac2127687bca4aab6ca0a953860f9de0b3f098/conf/config.level4.neon#L20-L30

File: src/Rules/DateTimeInstantiationRule.php [hunk 2]
```diff
@@ -0,0 +1,65 @@
...
+	/**
+	 * @param New_ $node
+	 */
+	public function processNode(Node $node, Scope $scope): array
+	{
+		if (
+			!($node->class instanceof \PhpParser\Node\Name)
+			|| !\in_array((string) $node->class, [DateTime::class, DateTimeImmutable::class], true)
+			|| \count($node->args) === 0
+			|| !($arg = $scope->getType($node->args[0]->value)) instanceof ConstantStringType
```
Comment (ondrejmirtes) [inline src/Rules/DateTimeInstantiationRule.php:32]: I don't like assignments inside if.

File: src/Rules/DateTimeInstantiationRule.php [hunk 3]
```diff
@@ -0,0 +1,65 @@
...
+	}
+
+	/**
+	 * @param New_ $node
+	 */
+	public function processNode(Node $node, Scope $scope): array
+	{
+		if (
+			!($node->class instanceof \PhpParser\Node\Name)
+			|| !\in_array((string) $node->class, [DateTime::class, DateTimeImmutable::class], true)
```
Comment (ondrejmirtes) [inline src/Rules/DateTimeInstantiationRule.php:30]: Please account for case-insensitive calls.

File: tests/PHPStan/Rules/DateTimeConstructorRuleTest.php [hunk 4]
```diff
@@ -0,0 +1,34 @@
...
+	{
+		return new DateTimeInstantiationRule();
+	}
+
+	public function test(): void
+	{
+		$this->analyse(
+			[__DIR__ . '/data/datetime-instantiation.php'],
+			[
+				['Instantiating DateTime with 2020.11.17 produces an error: Double time specification', 3],
```
Comment (ondrejmirtes) [inline tests/PHPStan/Rules/DateTimeConstructorRuleTest.php:21]: Please always do the multiline call.

**Review by ondrejmirtes — CHANGES_REQUESTED** [review]
And finally, please fix the build (Failures on PHP 8 are fine - they're on master already.)

File: tests/PHPStan/Rules/DateTimeConstructorRuleTest.php (hunk 4 above)
Comment (b1rdex) [reply tests/PHPStan/Rules/DateTimeConstructorRuleTest.php:21]: What do you mean?

Comment (ondrejmirtes) [reply tests/PHPStan/Rules/DateTimeConstructorRuleTest.php:21]:
```
[
    'Instantiating DateTime with 2020.11.17 produces an error: Double time specification',
    3,
],
```

File: src/Rules/DateTimeInstantiationRule.php [hunk 5]
```diff
@@ -0,0 +1,72 @@
...
+	 * @param New_ $node
+	 */
+	public function processNode(Node $node, Scope $scope): array
+	{
+		if (
+			!($node->class instanceof \PhpParser\Node\Name)
+			|| \count($node->args) === 0
+			|| (
+				strtolower((string) $node->class) !== strtolower(DateTime::class)
+				&& strtolower((string) $node->class) !== strtolower(DateTimeImmutable::class)
```
Comment (ondrejmirtes) [inline src/Rules/DateTimeInstantiationRule.php:33]: 1) This unnecessarily calls `strtolower((string) $node->class)` twice.
2) You can hardcode `datetime` instead.

**Review by ondrejmirtes — CHANGES_REQUESTED** [review]
This actually needs fixing:

```
Cannot access offset 'warnings' on array('warning_count' => int,     
         'warnings' => array<int, string>, 'error_count' => int, 'errors' =>  
         array<int, string>)|false.
```

**Comment by b1rdex** [comment]
Where that false comes from? I can’t reproduce returning false by that method. 

Отправлено с iPhone

**Comment by ondrejmirtes** [comment]
@b1rdex That's easy: https://3v4l.org/HBq0q

**Comment by ondrejmirtes** [comment]
It tells me that you haven't tested a case where the rule should not report anything :)

**Comment by b1rdex** [comment]
Actually there are tests with no errors. I haven’t tried calling it without constructing a date. Ok, will fix it tomorrow. 

Отправлено с iPhone

File: src/Rules/DateTimeInstantiationRule.php [hunk 6]
```diff
@@ -0,0 +1,70 @@
...
+
+		$errors = [];
+		$dateString = $arg->getValue();
+		try {
+			new DateTime($dateString);
+		} catch (\Throwable $e) {
+			// an exception is thrown for errors only but we want to catch warnings too
+		}
+		$lastErrors = DateTime::getLastErrors();
+		if ($lastErrors !== false) {
```
Comment (ondrejmirtes) [inline src/Rules/DateTimeInstantiationRule.php:48]: You can get rid of these errors by adding `|false` to the method in `resources/functionMap.php` :) And after that I'll be able to merge it.

**Review by ondrejmirtes — CHANGES_REQUESTED** [review]

File: src/Rules/DateTimeInstantiationRule.php (hunk 6 above)
Comment (b1rdex) [reply src/Rules/DateTimeInstantiationRule.php:48]: Would this help https://github.com/JetBrains/phpstorm-stubs/pull/963?

Comment (ondrejmirtes) [reply src/Rules/DateTimeInstantiationRule.php:48]: @b1rdex In some cases these stubs are used, but mostly it's a matter of these files: https://github.com/phpstan/phpstan-src/tree/master/resources

Comment (b1rdex) [reply src/Rules/DateTimeInstantiationRule.php:48]: I've updated functionMap

**Comment by ondrejmirtes** [comment]
Thank you!

**Comment by ondrejmirtes** [comment]
FYI I removed reporting warnings because it's not that useful: https://3v4l.org/TGsaG

**Pull request merged**